### DISCUSSION TITLE
lathe & udp

### DISCUSSION THREAD

**Discussion by jschoch** [multigcs/LinuxCNC-RIO#39] [Q&A]
I'd like to do a lathe setup with a W5500 native.  Can you let me know if this is doable and which branch to use and any oddities or thing I need to change/configure?

**Comment by jschoch** [comment]
Oh, i'm going to us a tang nano 9k

**Reply by multigcs** [reply]
at the moment i think the best is to use the main branch and the WT32-eth01 udp-bridge.

i still have network errors with the w5500, not many, but the udp-bridge is more stable.

for a lathe setup you can do everything the same way as for a milling machine, only the .hal/.ini files have to be adapted a little at the end (axis-gui lathe mode)

**Comment by jschoch** [comment]
I'm getting this error

```
PlatformIO/Projects/lincnc-rio/udp_testing/LinuxCNC-RIO$ python3 buildtool.py configs/Tangoboard/config-w5500.json 
loading board setup: TangNano9K
loading board setup: TangNano9K
to generate the flowchart, please install the 'graphviz' python module
generating files in Output/Tangoboard
generating gateware
generating linux-cnc component
generating linux-cnc config: axis
generating documentation
UNKNOWN PARAMETER (interface_w5500): 8'd168 <br/>
UNKNOWN PARAMETER (interface_w5500): 8'd10 <br/>
UNKNOWN PARAMETER (interface_w5500): 8'd193} <br/>
UNKNOWN PARAMETER (interface_w5500): 8'hAF <br/>
UNKNOWN PARAMETER (interface_w5500): 8'hFA <br/>
UNKNOWN PARAMETER (interface_w5500): 8'hCC <br/>
UNKNOWN PARAMETER (interface_w5500): 8'hE3 <br/>
UNKNOWN PARAMETER (interface_w5500): 8'h1C} <br/>
UNKNOWN PARAMETER (wiznet5500): 8'd168 <br/>
UNKNOWN PARAMETER (wiznet5500): 8'd10 <br/>
UNKNOWN PARAMETER (wiznet5500): 8'd193} <br/>
UNKNOWN PARAMETER (wiznet5500): 8'hAF <br/>
UNKNOWN PARAMETER (wiznet5500): 8'hFA <br/>
UNKNOWN PARAMETER (wiznet5500): 8'hCC <br/>
UNKNOWN PARAMETER (wiznet5500): 8'hE3 <br/>
UNKNOWN PARAMETER (wiznet5500): 8'h1C} <br/>
```

**Reply by multigcs** [reply]
ignore, only for the documentation module

**Comment by jschoch** [comment]
What linuxcnc version is required, I was in the rewrite branch due to the above error (before you answered that it was moot)
```
schoch@remoraPi4:~/linuxcnc-rio-udp/Component $ sudo halcompile --install ./rio.c 
[sudo] password for schoch: 
Compiling realtime rio.c
rio.c: In function ‘write_txbuffer’:
rio.c:1077:13: error: ‘data_t’ has no member named ‘MULTIPLEXER_OUTPUT_ID’
 1077 |     if (data->MULTIPLEXER_OUTPUT_ID < 0) {;
      |             ^~
rio.c:1078:13: error: ‘data_t’ has no member named ‘MULTIPLEXER_OUTPUT_ID’
 1078 |         data->MULTIPLEXER_OUTPUT_ID += 1;
      |             ^~
rio.c:1080:13: error: ‘data_t’ has no member named ‘MULTIPLEXER_OUTPUT_ID’
 1080 |         data->MULTIPLEXER_OUTPUT_ID = 0;
      |             ^~
make: *** [/usr/share/linuxcnc/Makefile.modinc:115: rio.o] Error 1

```

**Reply by jschoch** [reply]
I think this was because there was no rio.h or bcm2835.c  bcm2835.h

**Comment by jschoch** [comment]
I went back to main, but it isn't at all clear how what hal module talks over ethernet to the device, the rio.c seems to still be setup for direct SPI.

**Comment by jschoch** [comment]
Ok, so i added  the defines to rio.h

```
#define TRANSPORT_UDP
#define UDP_IP "192.168.10.194"
```

my config is this: 

```
"interface": [
    {
        "type": "w5500",
        "ip": "192.168.10.194",
        "port": "2390",
        "pins": {
            "MOSI": "75",
            "MISO": "74",
            "SCK": "76",
            "SEL": "77"
        }
    }
    ],
```

no idea if those pins would work, i tried to avoid the 1.8v pins.

But.. I can't ping.  Do you need a patch cable for it to work?  

Also, there are multiple different W5500 modules.  The one i'm using is also called USR-ES1.  

![image](https://github.com/multigcs/LinuxCNC-RIO/assets/20271/005351bb-5ebc-40a7-a8be-492a9cd8b668)


I have another variety that looks like this  
![image](https://github.com/multigcs/LinuxCNC-RIO/assets/20271/9afe3c2c-8bcb-4a00-aa84-1225e0e2fbef)

i wonder which you've tested with or how to troubleshoot?

**Comment by jschoch** [comment]
bad connection, it is working!

**Comment by jschoch** [comment]
both types of w5500 work

**Comment by multigcs** [comment]
you can add 
```
    "transport": "UDP",
    "ip": "192.168.10.194",
```
to the json-config to use UDP without editing the rio.h

**Reply by multigcs** [reply]
BTW, the rewrite branch is very bad, have to remove it, the real rewrite code is in another repository but better to use the main branch in this repo

**Comment by zayoo-HR** [comment]
Hello,
Do you still use this pins? I'm wondering is your connection ok on this pins? 
"interface": [
    {
        "type": "w5500",
        "ip": "192.168.10.194",
        "port": "2390",
        "pins": {
            "MOSI": "75",
            "MISO": "74",
            "SCK": "76",
            "SEL": "77"
        }
    }
    ],

This header pins also shares HDMI connector on TangNano9K board, which have RC component(see schematic).  And I'm afraid that will omit ISP communication.

**Reply by multigcs** [reply]
if it works it's good, if not you can simply remove the capacitor's with a screwdriver ;)

**Reply by jschoch** [reply]
these pins have been working for me.
